File: DiveIntoDeepLearning/Chapter06/TorchSharpChapter08/MLP.cs
using static TorchSharp.torch.nn;
using static TorchSharp.torch;
using static TorchSharp.torch.nn.functional;
using TorchSharp.Modules;

namespace TorchSharpChapter08;

public class MLP : Module<Tensor, Tensor>
{
    public MLP(string name) : base(name)
    {
        // LazyLinear is not yet available
        _hidden = Linear(20, 256);
        _output = Linear(256, 10);
    }

    public override Tensor forward(Tensor input) =>
        _output.call(relu(_hidden.call(input)));

    private readonly Linear _hidden;
    private readonly Linear _output;
}

File: DiveIntoDeepLearning/Chapter06/MLP.py
from torch import nn
from torch.nn import functional


class MLP(nn.Module):
    def __init__(self):
        super().__init__()

        self.hidden = nn.LazyLinear(256)  # define the hidden layer
        self.out = nn.LazyLinear(10)  # define the output layer

    # forward is the "forward propogation" function
    # returns the result of the model (output layer) in regards to X
    def forward(self, X):
        return self.out(functional.relu(self.hidden(X)))
